plugins/search-indexer-backend/src/types.ts

```diff
@@ -0,0 +1,97 @@
+/*
+ * Copyright 2021 Spotify AB
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * Base properties that all indexed documents must include, as well as some
+ * common properties that documents are encouraged to use where appropriate.
+ */
+export interface IndexableDocument {
+  /**
+   * The primary name of the document (e.g. name, title, identifier, etc).
+   */
+  title: string;
+
+  /**
+   * Free-form text of the document (e.g. description, content, etc).
+   */
+  text: string;
+
+  /**
+   * The relative or absolute URL of the document (target when a search result
+   * is clicked).
+   */
+  location: string;
```

Comment:
Not sure this could actually be controlled from here tbh, may need to make it more of an identifier that gets parsed in the frontend and converted to a link there

Reply:
I'd _really_ prefer to keep that out of the frontend.  It helps keep the frontend a much "dumber," generic display layer, but also treating the URL as de facto identifier simplifies search quality control.

e.g. when performing a query and checking for known, good results and their rank, it's much easier to reason about a URL (which has implicit hierarchy and other metadata), rather than to have to consider different types of results on a case by case basis.

Not something I imagine will be a huge concern for early Backstage adopters, but it's been a pain-point internally.

Is this hard because we allow you to stick entities wherever you want in the frontend?

Reply:
> Is this hard because we allow you to stick entities wherever you want in the frontend?

Yep, and all other kind of pages that you might want to link to. The frontend app is the source of truth for routing those. I think the two main options here would be to either have a mapping of `result -> url` in the frontend, or a mapping of `result -> url` in the backend, where if it's in the backend it would have to duplicate the routing configuration of the frontend.

Reply:
Ah, the context shifts necessary when coming from an everything-is-everywhere background.

What if we had a configurable location string template that you just configured per index-type?  e.g. we ship this by default:

```yaml
searchLocationTemplates:
  softwareCatalog: '/{catalog/{namespace}/{type}/{name}'
  techDocs: '/docs/{namespace}/{type}/{name}/{pagePath}'
```

And then if you pick different routes on the frontend, you update the patterns in the backend config?

Reply:
> Ah, the context shifts necessary when coming from an everything-is-everywhere background.
> 
> What if we had a configurable location string template that you just configured per index-type? e.g. we ship this by default:
> 
> ```yaml
> searchLocationTemplates:
>   softwareCatalog: '/catalog/{namespace}/{type}/{name}'
>   techDocs: '/docs/{namespace}/{type}/{name}/{pagePath}'
> ```
> 
> And then if you pick different routes on the frontend, you update the patterns in the backend config?

Yep that'd be one way to do it, although given that structure, could we instead represent the location as the following instead of a URL?

```json
{
  "index": "techDocs",
  "params": {
    "namespace": "...",
    "type": "...",
    "name": "...",
    "pagePath": "..."
  }
}
```

Reply:
Sure, but then you expand to indexing stuff like...  StackOverflow Enterprise Questions or BigQuery datasets or...  Anything that might not map well onto to the software catalog's metadata model, and it starts breaking down.

The lowest common denominator of a searchable thing is that it exists somewhere, and when you click on a result for it, you want to be taken there.

Reply:
Absolute URLs that link outside of Backstage could ofc be returned as a full URL. The techdocs page in your example isn't a clean mapping to the catalog, but I think our examples above would work just fine. My point is that you're unlikely to get away from having to have configurable paths that get reconstructed somehow, and it's likely that sending them over in parameter form to the frontend provides a much better DX as it avoids fragmentation of the configuration

Reply:
I guess I'm thinking of DX from a different perspective: the cost here would be 1-time duplication of config, but the benefit is not having to duplicate the _logic_ to identify an indexed document in other contexts.

As someone responsible for search in Backstage, tuning queries to get the best results, I might need to be able to assert the rank of a particular document for a given query term as I'm making changes to the compiled query.  And I'd need to do that generically across document types.

```json
[{
  "givenQuery": "CI/CD",
  "resultRankShouldBeLessThan": 5,
  "ofExpectedDocument": "(a TechDoc page about the build system)"
}, {
  "givenQuery": "CI/CD",
  "resultRankShouldBeLessThan": 5,
  "ofExpectedDocument": "(the catalog entity for the build system)"
}]
```

^ The context we'd want to make these assertions is in the context of where the search engine query is actually compiled, which will be in the backend.

Or imagine a scenario where you want to monitor the results for a given set of queries over time (to see how changes to the index as well as changes to queries affect the rank of well-searched things over time).  You might set up a completely separate service or pipeline that generates such a dataset.  And you'd end up having to duplicate the document identification logic to populate a `document` column in the dataset.

Reply:
Yeah it's true that it would be bad to have only have the logic to create useful results in the frontend. It would make it tricky to consume the search API from other sources, like if we add a  CLI in the future. If we make it config driven we could potentially do a one-way validation and warn in the frontend if the paths in config don't match the registered routes in the frontend.